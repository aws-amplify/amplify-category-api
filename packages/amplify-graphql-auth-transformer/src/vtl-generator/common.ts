import { and, bool, equals, Expression, methodCall, qref, ref, str } from 'graphql-mapping-template';
import { IAM_AUTH_TYPE } from '../utils';

// note in the resolver that operation is protected by auth
export const setHasAuthExpression: Expression = qref(methodCall(ref('ctx.stash.put'), str('hasAuth'), bool(true)));

// a logical condition which returns true if request is made with non-Cognito IAM principal
export const isNonCognitoIAMPrincipal = and([
  equals(ref('util.authType()'), str(IAM_AUTH_TYPE)),
  methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityPoolId')),
  methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityId')),
]);
