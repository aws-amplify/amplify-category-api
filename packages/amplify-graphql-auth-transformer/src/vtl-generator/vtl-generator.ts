import { TransformerContextProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { FieldDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { aws_dynamodb as dynamodb } from 'aws-cdk-lib';
import { ConfiguredAuthProviders, RelationalPrimaryMapConfig, RoleDefinition } from '../utils';

export interface AuthVTLGenerator {
  generateAuthExpressionForCreate: (
    ctx: TransformerContextProvider,
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
  ) => string;

  generateAuthExpressionForUpdate: (
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
  ) => string;

  generateAuthRequestExpression: (ctx: TransformerContextProvider, def: ObjectTypeDefinitionNode) => string;

  generateAuthExpressionForDelete: (
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
  ) => string;

  generateAuthExpressionForField: (
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
    fieldName: string | undefined, // Default 'undefined'
  ) => string;

  generateFieldAuthResponse: (operation: string, fieldName: string, subscriptionsEnabled: boolean) => string;

  generateAuthExpressionForQueries: (
    ctx: TransformerContextProvider,
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
    def: ObjectTypeDefinitionNode,
    indexName: string | undefined, // Default 'undefined'
  ) => string;

  generateAuthExpressionForSearchQueries: (
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
    allowedAggFields: Array<string>,
  ) => string;

  generateAuthExpressionForSubscriptions: (providers: ConfiguredAuthProviders, roles: Array<RoleDefinition>) => string;

  setDeniedFieldFlag: (operation: string, subscriptionsEnabled: boolean) => string;

  generateAuthExpressionForRelationQuery: (
    ctx: TransformerContextProvider,
    def: ObjectTypeDefinitionNode,
    field: FieldDefinitionNode,
    relatedModelObject: ObjectTypeDefinitionNode,
    providers: ConfiguredAuthProviders,
    roles: Array<RoleDefinition>,
    fields: ReadonlyArray<FieldDefinitionNode>,
  ) => string;

  generateFieldResolverForOwner: (entity: string) => string;

  generateSandboxExpressionForField: (sandboxEnabled: boolean, genericIamAccessEnabled: boolean) => string;
}
