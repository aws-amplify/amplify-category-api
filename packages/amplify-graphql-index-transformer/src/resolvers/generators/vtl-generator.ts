import { TransformerContextProvider, TransformerResolverProvider } from '@aws-amplify/graphql-transformer-interfaces';

import { IndexDirectiveConfiguration, PrimaryKeyDirectiveConfiguration } from '../../types';

export interface IndexVTLGenerator {
  generatePrimaryKeyVTL(
    config: PrimaryKeyDirectiveConfiguration,
    ctx: TransformerContextProvider,
    resolverMap: Map<TransformerResolverProvider, string>,
  ): void;
  generateIndexQueryRequestTemplate(
    config: IndexDirectiveConfiguration,
    ctx: TransformerContextProvider,
    tableName: string,
    operationName: string,
  ): string;
}
