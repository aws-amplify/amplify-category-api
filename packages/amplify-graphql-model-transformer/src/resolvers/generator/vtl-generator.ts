import { ModelDirectiveConfiguration } from "../../directive";

export interface OperationConfig {
  operation?: string;
  operationName?: string;
}

export interface ModelVTLGenerator {
  generateUpdateRequestTemplate(modelName: string, isSyncEnabled: boolean, config?: OperationConfig): string;
  generateCreateRequestTemplate(modelName: string, modelIndexFields: string[], config?: OperationConfig): string;
  generateCreateInitSlotTemplate(modelConfig: ModelDirectiveConfiguration, config?: OperationConfig): string;
  generateDeleteRequestTemplate(modelName: string, isSyncEnabled: boolean, config?: OperationConfig): string;
  generateUpdateInitSlotTemplate(modelConfig: ModelDirectiveConfiguration, config?: OperationConfig): string;
  generateGetRequestTemplate(config?: OperationConfig): string;
  generateGetResponseTemplate(isSyncEnabled: boolean, config?: OperationConfig): string;
  generateListRequestTemplate(config?: OperationConfig): string;
  generateSyncRequestTemplate(config?: OperationConfig): string;
  generateSubscriptionRequestTemplate(config?: OperationConfig): string;
  generateSubscriptionResponseTemplate(config?: OperationConfig): string;
  generateDefaultResponseMappingTemplate(isSyncEnabled: boolean, mutation: boolean): string;
}
