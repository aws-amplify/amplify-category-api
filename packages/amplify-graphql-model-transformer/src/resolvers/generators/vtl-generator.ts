import { ModelDirectiveConfiguration } from '../../directive';

export type ModelRequestConfig = {
  modelName: string;
  operation: string;
  operationName: string;
};

export type ModelUpdateRequestConfig = ModelRequestConfig & {
  modelIndexFields: string[];
  isSyncEnabled: boolean;
};

export type ModelDeleteRequestConfig = ModelUpdateRequestConfig;

export type ModelCreateRequestConfig = ModelRequestConfig & {
  modelIndexFields: string[];
};

export type ModelCreateInitSlotConfig = {
  modelConfig: ModelDirectiveConfiguration;
};

export type ModelUpdateInitSlotConfig = ModelCreateInitSlotConfig;

export type ModelGetResponseConfig = ModelUpdateRequestConfig;

export type ModelDefaultResponseConfig = ModelRequestConfig & {
  isSyncEnabled: boolean;
  mutation: boolean;
};

export interface ModelVTLGenerator {
  generateUpdateRequestTemplate(config: ModelUpdateRequestConfig): string;
  generateCreateRequestTemplate(config: ModelCreateRequestConfig): string;
  generateCreateInitSlotTemplate(config: ModelCreateInitSlotConfig): string;
  generateDeleteRequestTemplate(config: ModelDeleteRequestConfig): string;
  generateUpdateInitSlotTemplate(config: ModelUpdateInitSlotConfig): string;
  generateGetRequestTemplate(config: ModelRequestConfig): string;
  generateGetResponseTemplate(config: ModelGetResponseConfig): string;
  generateListRequestTemplate(config: ModelRequestConfig): string;
  generateSyncRequestTemplate(config: ModelRequestConfig): string;
  generateSubscriptionRequestTemplate(): string;
  generateSubscriptionResponseTemplate(): string;
  generateDefaultResponseMappingTemplate(config: ModelDefaultResponseConfig): string;
}
