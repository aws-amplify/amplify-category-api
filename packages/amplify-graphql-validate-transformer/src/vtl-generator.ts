import { printBlock, raw } from 'graphql-mapping-template';
import { ValidationRuleConfig, ValidationsByField } from './types';

/**
 * Generates a combined VTL snippet for all validations in a type.
 * @param typeName - The name of the type to generate validations for
 * @param validationsByField - Map of field names to their validation configurations
 * @returns Combined VTL snippet containing all validations for the type
 */
export const generateTypeValidationSnippet = (typeName: string, validationsByField: ValidationsByField): string => {
  const fieldValidationBlocks: string[] = [];

  for (const [fieldName, validations] of Object.entries(validationsByField)) {
    const fieldBlock = generateFieldValidationSnippet(fieldName, validations);
    fieldValidationBlocks.push(fieldBlock);
  }

  // Combine all field validations into a single VTL snippet
  const combinedSnippet = printBlock(`Validations for type ${typeName} **`)(raw(fieldValidationBlocks.join('\n')));
  return combinedSnippet;
};

/**
 * Generates a VTL snippet for all validations for a field.
 * @param fieldName - The name of the field to validate
 * @param validations - Array of validation configurations for the field
 * @returns A VTL code block that performs all validations for the field
 */
export const generateFieldValidationSnippet = (fieldName: string, validations: ValidationRuleConfig[]): string => {
  const validationLines = [`#if( !$util.isNull($ctx.args.input.${fieldName}) )`];

  // Add each validation with its error check
  for (const validation of validations) {
    const { validationType, validationValue, errorMessage } = validation;
    const escapedValue = escapeSingleQuotes(validationValue);
    const escapedErrorMessage = escapeSingleQuotes(errorMessage);
    const validationVar = `${validationType}ValidationPassed`;
    const validationCheck = getValidationCheck(fieldName, validationVar, validationType, escapedValue);

    validationLines.push(`  ${validationCheck}`, `  #if(!$${validationVar})`, `    $util.error('${escapedErrorMessage}')`, '  #end');
  }

  validationLines.push('#end');

  return printBlock(`Validations for field ${fieldName}`)(raw(validationLines.join('\n')));
};

/**
 * Generates a VTL snippet for a validation rule.
 * @param fieldName - The name of the field to validate
 * @param validationVar - The name of the validation variable
 * @param validationType - The type of validation to perform
 * @param value - The value to compare against
 * @returns A VTL code block that performs the validation check
 */
const getValidationCheck = (fieldName: string, validationVar: string, validationType: string, value: string): string => {
  const fieldRef = `$ctx.args.input.${fieldName}`;

  switch (validationType.toLowerCase()) {
    case 'gt':
      return `#set($${validationVar} = ${fieldRef} > ${value})`;
    case 'lt':
      return `#set($${validationVar} = ${fieldRef} < ${value})`;
    case 'gte':
      return `#set($${validationVar} = ${fieldRef} >= ${value})`;
    case 'lte':
      return `#set($${validationVar} = ${fieldRef} <= ${value})`;
    case 'minlength':
      return `#set($${validationVar} = ${fieldRef}.length() >= ${value})`;
    case 'maxlength':
      return `#set($${validationVar} = ${fieldRef}.length() <= ${value})`;
    case 'startswith':
      return `#set($${validationVar} = ${fieldRef}.startsWith('${value}'))`;
    case 'endswith':
      return `#set($${validationVar} = ${fieldRef}.endsWith('${value}'))`;
    case 'matches':
      return `#set($${validationVar} = $util.matches('${value}', ${fieldRef}))`;
    default:
      throw new Error(`Unsupported validation type: ${validationType}`);
  }
};

/**
 * Escapes single quotes in a string for use in VTL templates
 * @param str - The string to escape
 * @returns The escaped string
 */
const escapeSingleQuotes = (str: string): string => {
  return str.replace(/'/g, "''");
};
