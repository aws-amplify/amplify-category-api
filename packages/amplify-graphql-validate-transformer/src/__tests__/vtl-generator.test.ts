import { makeValidationSnippet } from '../vtl-generator';

describe('vtl-generator', () => {
  describe('Test `makeValidationSnippet` with different validation types', () => {
    const testCases = [
      {
        fieldName: 'age',
        validationType: 'gt',
        validationValue: '18',
        errorMessage: 'Age must be greater than 18',
      },
      {
        fieldName: 'price',
        validationType: 'lt',
        validationValue: '1000',
        errorMessage: 'Price must be less than 1000',
      },
      {
        fieldName: 'quantity',
        validationType: 'gte',
        validationValue: '0',
        errorMessage: 'Quantity must be non-negative',
      },
      {
        fieldName: 'rating',
        validationType: 'lte',
        validationValue: '5',
        errorMessage: 'Rating must be at most 5',
      },
      {
        fieldName: 'name',
        validationType: 'minlength',
        validationValue: '3',
        errorMessage: 'Name must be at least 3 characters',
      },
      {
        fieldName: 'bio',
        validationType: 'maxlength',
        validationValue: '500',
        errorMessage: 'Bio must not exceed 500 characters',
      },
      {
        fieldName: 'url',
        validationType: 'startswith',
        validationValue: 'https://',
        errorMessage: 'URL must start with https://',
      },
      {
        fieldName: 'filename',
        validationType: 'endswith',
        validationValue: '.pdf',
        errorMessage: 'File must be a PDF',
      },
      {
        fieldName: 'email',
        validationType: 'matches',
        validationValue: '^[A-Za-z0-9+_.-]+@(.+)$',
        errorMessage: 'Invalid email format',
      },
    ];

    testCases.forEach(({ fieldName, validationType, validationValue, errorMessage }) => {
      it(`generates correct VTL for ${validationType} validation`, () => {
        const result = makeValidationSnippet(fieldName, validationType, validationValue, errorMessage);
        expect(result).toMatchSnapshot(`${fieldName} ${validationType} validation`);
      });
    });

    it('throws error for unknown validation types', () => {
      expect(() => makeValidationSnippet('field', 'unknown', 'value', 'error')).toThrow('Unsupported validation type: unknown');
    });

    it('includes validation details in block comment', () => {
      const result = makeValidationSnippet('age', 'gt', '18', 'error');
      expect(result).toMatchSnapshot('validation details in block comment');
    });
  });
});
