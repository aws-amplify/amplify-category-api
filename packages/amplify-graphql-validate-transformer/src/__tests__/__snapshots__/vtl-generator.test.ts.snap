// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for endswith validation: filename endswith validation 1`] = `
"## [Start] Validating \\"filename\\" with type \\"endswith\\" and value \\".pdf\\". **
#if( !$util.isNull($ctx.args.input.filename) )
  #set($validationPassed = $ctx.args.input.filename.endsWith('.pdf'))
  #if(!$validationPassed)
    $util.error('File must be a PDF')
  #end
#end
## [End] Validating \\"filename\\" with type \\"endswith\\" and value \\".pdf\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for gt validation: age gt validation 1`] = `
"## [Start] Validating \\"age\\" with type \\"gt\\" and value \\"18\\". **
#if( !$util.isNull($ctx.args.input.age) )
  #set($validationPassed = $ctx.args.input.age > 18)
  #if(!$validationPassed)
    $util.error('Age must be greater than 18')
  #end
#end
## [End] Validating \\"age\\" with type \\"gt\\" and value \\"18\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for gte validation: quantity gte validation 1`] = `
"## [Start] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.quantity) )
  #set($validationPassed = $ctx.args.input.quantity >= 0)
  #if(!$validationPassed)
    $util.error('Quantity must be non-negative')
  #end
#end
## [End] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for lt validation: price lt validation 1`] = `
"## [Start] Validating \\"price\\" with type \\"lt\\" and value \\"1000\\". **
#if( !$util.isNull($ctx.args.input.price) )
  #set($validationPassed = $ctx.args.input.price < 1000)
  #if(!$validationPassed)
    $util.error('Price must be less than 1000')
  #end
#end
## [End] Validating \\"price\\" with type \\"lt\\" and value \\"1000\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for lte validation: rating lte validation 1`] = `
"## [Start] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **
#if( !$util.isNull($ctx.args.input.rating) )
  #set($validationPassed = $ctx.args.input.rating <= 5)
  #if(!$validationPassed)
    $util.error('Rating must be at most 5')
  #end
#end
## [End] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for matches validation: email matches validation 1`] = `
"## [Start] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **
#if( !$util.isNull($ctx.args.input.email) )
  #set($validationPassed = $util.matches('^[A-Za-z0-9+_.-]+@(.+)$', $ctx.args.input.email))
  #if(!$validationPassed)
    $util.error('Invalid email format')
  #end
#end
## [End] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for maxlength validation: bio maxlength validation 1`] = `
"## [Start] Validating \\"bio\\" with type \\"maxlength\\" and value \\"500\\". **
#if( !$util.isNull($ctx.args.input.bio) )
  #set($validationPassed = $ctx.args.input.bio.length() <= 500)
  #if(!$validationPassed)
    $util.error('Bio must not exceed 500 characters')
  #end
#end
## [End] Validating \\"bio\\" with type \\"maxlength\\" and value \\"500\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for minlength validation: name minlength validation 1`] = `
"## [Start] Validating \\"name\\" with type \\"minlength\\" and value \\"3\\". **
#if( !$util.isNull($ctx.args.input.name) )
  #set($validationPassed = $ctx.args.input.name.length() >= 3)
  #if(!$validationPassed)
    $util.error('Name must be at least 3 characters')
  #end
#end
## [End] Validating \\"name\\" with type \\"minlength\\" and value \\"3\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types generates correct VTL for startswith validation: url startswith validation 1`] = `
"## [Start] Validating \\"url\\" with type \\"startswith\\" and value \\"https://\\". **
#if( !$util.isNull($ctx.args.input.url) )
  #set($validationPassed = $ctx.args.input.url.startsWith('https://'))
  #if(!$validationPassed)
    $util.error('URL must start with https://')
  #end
#end
## [End] Validating \\"url\\" with type \\"startswith\\" and value \\"https://\\". **"
`;

exports[`vtl-generator Test \`makeValidationSnippet\` with different validation types includes validation details in block comment: validation details in block comment 1`] = `
"## [Start] Validating \\"age\\" with type \\"gt\\" and value \\"18\\". **
#if( !$util.isNull($ctx.args.input.age) )
  #set($validationPassed = $ctx.args.input.age > 18)
  #if(!$validationPassed)
    $util.error('error')
  #end
#end
## [End] Validating \\"age\\" with type \\"gt\\" and value \\"18\\". **"
`;
